### PR TITLE
client resync

### LINKED ISSUE #2328: Add support for Full Sync Client Resync

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?
Add support for client resync, and it's three different modes.

This closes #2328.

## ☑️ ToDos

* [x] 📝 Changelog entry
* ~[ ] 📝 `Compatibility` label is updated or copied from previous entry~
* [x] 🚦 Tests
* ~[ ] 📝 Public documentation PR created or is not necessary~
* ~[ ] 💥 `Breaking` label has been applied or is not necessary~

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* ~[ ] Chrome debug API is updated if API is available on React Native~

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,21 @@
+x.x.x Release notes (yyyy-MM-dd)
+=============================================================
+### Enhancements
+* Added support for Client Resync which automatically will recover the local Realm in case the server is rolled back. This largely replaces the Client Reset mechanism. Can be configured using `Realm.Sync.Configuration.clientResyncMode`. Three modes are available: `'recover'`, `'discard`', and `'manual'` but currently only `'manual'` is supported for query-based sync. Default is `'manual'`. ([#2328](https://github.com/realm/realm-js/issues/2328))
```
Comment (bmunkholm) [inline CHANGELOG.md]: isn't default value "recover"?

File: scripts/changelog-header.sh [hunk 2]
```diff
@@ -14,7 +14,7 @@ x.x.x Release notes (yyyy-MM-dd)
 * None.
 
 ### Compatibility
-* Realm Object Server: 3.21.0 or later.
+* Realm Object Server: 3.23.1 or later.
```
Comment (bmunkholm) [inline scripts/changelog-header.sh]: @kneth I took the liberty to add this, so we don't have to remember to fix this at the next releases :-)

File: src/js_sync.hpp [hunk 3]
```diff
@@ -1254,8 +1254,24 @@ void SyncClass<T>::populate_sync_config(ContextType ctx, ObjectType realm_constr
         }
 #endif
 
-    // FIXME: remove once Client Resync is implemented
-    config.sync_config->client_resync_mode = realm::ClientResyncMode::Manual;
+        ClientResyncMode clientResyncMode = realm::ClientResyncMode::Manual;
```
Comment (bmunkholm) [inline src/js_sync.hpp]: Shouldn't the default be recover?

File: CHANGELOG.md (hunk 1 above)
Comment (kneth) [reply CHANGELOG.md]: If `recover` is default, it is a breaking change (`manual` is like the old client reset). I'll update the doc.

File: src/js_sync.hpp (hunk 3 above)
Comment (kneth) [reply src/js_sync.hpp]: See above.

File: CHANGELOG.md (hunk 1 above)
Comment (bmunkholm) [reply CHANGELOG.md]: Is that really breaking? It should be transparent? Java also has Recover as default?

Comment (bmunkholm) [reply CHANGELOG.md]: But it seems Cocoa has Manual as the default?

**Pull request merged**